core/wasm-interface/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+	fn write_memory(&mut self, address: Pointer<u8>, data: &[u8]) -> Result<(), String>;
+	/// Allocate a memory instance of `size` bytes.
+	fn allocate_memory(&mut self, size: WordSize) -> Result<Pointer<u8>, String>;
+	/// Deallocate a given memory instance.
+	fn deallocate_memory(&mut self, ptr: Pointer<u8>) -> Result<(), String>;
+}
+
+/// Something that provides implementations for host functions.
+pub trait HostFunctions {
+	/// Try to resolve the function with the given name and signature.
+	fn resolve_function(name: &str, signature: &Signature) -> Result<Option<FunctionRef>, String>;
+
+	/// Returns the number of host functions this type provides.
+	fn function_count() -> usize;
+
+	/// Execute the function at the given index.
+	///
+	/// - `index` - Is equal to the index given to `FunctionRef`.
+	/// - `args` - The arguments given to the function.
+	/// - `context` - Provides access to the allocator and memory of the memory wasm instance.
+	fn execute_function<A: Iterator<Item=Value>>(
+		index: usize,
+		args: A,
+		context: &mut dyn HostFunctionsContext,
+	) -> Result<Option<Value>, String>;
+}
+
+/// Something that provides implementations for inherent host functions.
+///
+/// Will be used interanlly by Substrate to provide functions like `malloc`, `free` etc.
+pub trait InherentHostFunctions {
+	/// Try to resolve the function with the given name and signature.
+	fn resolve_function(name: &str, signature: &Signature) -> Result<Option<FunctionRef>, String>;
+
+	/// Returns the number of host functions this type provides.
+	fn function_count() -> usize;
+
+	/// Execute the function at the given index.
+	///
+	/// - `index` - Is equal to the index given to `FunctionRef`.
+	/// - `args` - The arguments given to the function.
+	fn execute_function<A: Iterator<Item=Value>>(
+		&mut self,
+		index: usize,
+		args: A,
+	) -> Result<Option<Value>, String>;
+}
+
+/// Something that can be converted into a wasm compatible `Value`.
+pub trait IntoValue {
+	/// The type of the value in wasm.
+	const VALUE_TYPE: ValueType;
+
+	/// Convert `self` into a wasm `Value`.
+	fn into_value(self) -> Value;
+}
+
+/// Something that can may be created from a wasm `Value`.
+pub trait TryFromValue: Sized {
+	/// Try to convert the given `Value` into `Self`.
+	fn try_from_value(val: Value) -> Option<Self>;
+}
+
+macro_rules! impl_into_and_from_value {
+	(
+		$(
+			$type:ty, $( < $gen:ident >, )? $value_variant:ident,
+		)*
+	) => {
+		$(
+			impl $( <$gen> )? IntoValue for $type {
+				const VALUE_TYPE: ValueType = ValueType::$value_variant;
+				fn into_value(self) -> Value { Value::$value_variant(self as _) }
+			}
+
+			impl $( <$gen> )? TryFromValue for $type {
+				fn try_from_value(val: Value) -> Option<Self> {
+					match val {
+						Value::$value_variant(val) => Some(val as _),
+						_ => None,
+					}
+				}
+			}
+		)*
+	}
+}
+
+impl_into_and_from_value! {
+	u32, I32,
+	i32, I32,
+	u64, I64,
+	i64, I64,
+	*const T, <T>, I32,
```

Comment:
Is this and the below still necessary now that we have `Pointer<T>`?

Reply:
Good point.